Add `NoChunk` wrapper for pipeline args.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57325 Add `NoChunk` wrapper for pipeline args.**
* #57226 Support non-Tensor args in the Pipe API

As per the design outlined in
https://github.com/pytorch/pytorch/issues/53952, adding a `NoChunk` wrapper for
pipeline parallelism inputs.

If a Tensor is wrapped with this wrapper, the pipeline implementation does not
split this Tensor across micro-batches and instead just replicates this tensor
as-is similar to non-tensors.

Differential Revision: [D28109277](https://our.internmc.facebook.com/intern/diff/D28109277/)

pritamdamania committed:
Add `NoChunk` wrapper for pipeline args.

As per the design outlined in
https://github.com/pytorch/pytorch/issues/53952, adding a `NoChunk` wrapper for
pipeline parallelism inputs.

If a Tensor is wrapped with this wrapper, the pipeline implementation does not
split this Tensor across micro-batches and instead just replicates this tensor
as-is similar to non-tensors.

Differential Revision: [D28109277](https://our.internmc.facebook.com/intern/diff/D28109277/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add `NoChunk` wrapper for pipeline args."


As per the design outlined in
https://github.com/pytorch/pytorch/issues/53952, adding a `NoChunk` wrapper for
pipeline parallelism inputs.

If a Tensor is wrapped with this wrapper, the pipeline implementation does not
split this Tensor across micro-batches and instead just replicates this tensor
as-is similar to non-tensors.

Differential Revision: [D28109277](https://our.internmc.facebook.com/intern/diff/D28109277/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add `NoChunk` wrapper for pipeline args."


As per the design outlined in
https://github.com/pytorch/pytorch/issues/53952, adding a `NoChunk` wrapper for
pipeline parallelism inputs.

If a Tensor is wrapped with this wrapper, the pipeline implementation does not
split this Tensor across micro-batches and instead just replicates this tensor
as-is similar to non-tensors.

Differential Revision: [D28109277](https://our.internmc.facebook.com/intern/diff/D28109277/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add `NoChunk` wrapper for pipeline args."


As per the design outlined in
https://github.com/pytorch/pytorch/issues/53952, adding a `NoChunk` wrapper for
pipeline parallelism inputs.

If a Tensor is wrapped with this wrapper, the pipeline implementation does not
split this Tensor across micro-batches and instead just replicates this tensor
as-is similar to non-tensors.

Differential Revision: [D28109277](https://our.internmc.facebook.com/intern/diff/D28109277/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add `NoChunk` wrapper for pipeline args."


As per the design outlined in
https://github.com/pytorch/pytorch/issues/53952, adding a `NoChunk` wrapper for
pipeline parallelism inputs.

If a Tensor is wrapped with this wrapper, the pipeline implementation does not
split this Tensor across micro-batches and instead just replicates this tensor
as-is similar to non-tensors.

Differential Revision: [D28109277](https://our.internmc.facebook.com/intern/diff/D28109277/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add `NoChunk` wrapper for pipeline args."


As per the design outlined in
https://github.com/pytorch/pytorch/issues/53952, adding a `NoChunk` wrapper for
pipeline parallelism inputs.

If a Tensor is wrapped with this wrapper, the pipeline implementation does not
split this Tensor across micro-batches and instead just replicates this tensor
as-is similar to non-tensors.

Differential Revision: [D28109277](https://our.internmc.facebook.com/intern/diff/D28109277/)

[ghstack-poisoned]

wayi1 reviewed (approved):
LGTM!

Can we clarify that the replicated `NoChunk` tensors still share the underlying data storage, and this wrapper will not cause any deep copies?

rohan-varma commented on torch/distributed/pipeline/sync/microbatch.py:
```diff
@@ -173,7 +190,11 @@ def scatter(*inputs, chunks: int) -> List[Batch]:
             # Replicate non-tensors.
             for i in range(chunks):
-                batches[i].append(input)
+                if isinstance(input, NoChunk):
```
nit: update the comment on 191 to say replicate non-tensors, or tensors that have been marked with `NoChunk`?

pritamdamania87 commented:
> Can we clarify that the replicated NoChunk tensors still share the underlying data storage, and this wrapper will not cause any deep copies?

`a = b` for tensors in python doesn't create any deep copies in PyTorch, it is just refcounting under the hood. There is nothing different happening here compared to usual PyTorch tensor operations that users already know.

pritamdamania committed:
Update on "Add `NoChunk` wrapper for pipeline args."


As per the design outlined in
https://github.com/pytorch/pytorch/issues/53952, adding a `NoChunk` wrapper for
pipeline parallelism inputs.

If a Tensor is wrapped with this wrapper, the pipeline implementation does not
split this Tensor across micro-batches and instead just replicates this tensor
as-is similar to non-tensors.

Differential Revision: [D28109277](https://our.internmc.facebook.com/intern/diff/D28109277/)

[ghstack-poisoned]
